File: companion/app/src/main/java/com/example/sentrykey/SyncCrypto.kt
package com.example.sentrykey

import android.util.Base64
import java.security.SecureRandom
import javax.crypto.Mac
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Optional passphrase encryption for the phone <-> watch BLE sync string.
 *
 * Why a custom HMAC-SHA1 construction instead of AES: the watch must be able to
 * decrypt, and the Garmin watch's native crypto (Toybox.Cryptography) faults
 * uncatchably for SHA-1 on some fenix 8 firmware, so the watch only has a pure
 * Monkey C HMAC-SHA1. This scheme is built entirely from HMAC-SHA1 so all three
 * platforms (Android/iOS/watch) can implement it from the same primitive.
 *
 * Construction (everything HMAC-SHA1 based):
 *   salt   = 16 random bytes
 *   nonce  =  8 random bytes
 *   key    = PBKDF2-HMAC-SHA1(passphrase, salt, ITERATIONS, dkLen=40)
 *   encKey = key[0..20), macKey = key[20..40)
 *   keystream block i = HMAC-SHA1(encKey, nonce || be32(i))   // 20 bytes each
 *   ciphertext = plaintext XOR keystream[0 .. plaintext.size)
 *   mac    = HMAC-SHA1(macKey, salt || nonce || ciphertext)   // 20 bytes
 *   wire   = "SKENC1:" + base64( salt(16) || nonce(8) || mac(20) || ciphertext )
 *
 * This is defense-in-depth over a BLE link that is already encrypted at the link
 * layer; it is deliberately simple and is NOT a security-audited protocol.
 *
 * NOTE: ITERATIONS is kept low because the watch runs PBKDF2 in interpreted
 * Monkey C. Raising it strengthens the KDF but slows watch sync proportionally.
 */
object SyncCrypto {
    const val MARKER = "SKENC1:"
    const val ITERATIONS = 1000

    private const val SALT_LEN = 16
    private const val NONCE_LEN = 8
    private const val MAC_LEN = 20
    private const val HMAC = "HmacSHA1"

    fun isEncrypted(payload: String): Boolean = payload.startsWith(MARKER)

    fun encrypt(plaintext: String, passphrase: String): String {
        val random = SecureRandom()
        val salt = ByteArray(SALT_LEN).also { random.nextBytes(it) }
        val nonce = ByteArray(NONCE_LEN).also { random.nextBytes(it) }
        return encryptWith(plaintext, passphrase, salt, nonce)
    }

    // Deterministic core, separated so a test can pin a fixed salt/nonce vector.
    internal fun encryptWith(plaintext: String, passphrase: String, salt: ByteArray, nonce: ByteArray): String {
        val (encKey, macKey) = deriveKeys(passphrase, salt)
        val plainBytes = plaintext.toByteArray(Charsets.UTF_8)
        val ciphertext = xorKeystream(plainBytes, encKey, nonce)
        val mac = hmac(macKey, salt + nonce + ciphertext)
        val blob = salt + nonce + mac + ciphertext
        return MARKER + Base64.encodeToString(blob, Base64.NO_WRAP)
    }

    /** Throws [BadPasswordException] if the MAC fails (wrong passphrase or tamper). */
    fun decrypt(payload: String, passphrase: String): String {
        if (!isEncrypted(payload)) throw BadPasswordException("Not an encrypted sync payload.")
        val blob = try {
            Base64.decode(payload.substring(MARKER.length), Base64.NO_WRAP)
        } catch (e: Exception) {
            throw BadPasswordException("Corrupt sync payload.")
        }
        val minLen = SALT_LEN + NONCE_LEN + MAC_LEN
        if (blob.size < minLen) throw BadPasswordException("Corrupt sync payload.")

        val salt = blob.copyOfRange(0, SALT_LEN)
        val nonce = blob.copyOfRange(SALT_LEN, SALT_LEN + NONCE_LEN)
        val mac = blob.copyOfRange(SALT_LEN + NONCE_LEN, minLen)
        val ciphertext = blob.copyOfRange(minLen, blob.size)

        val (encKey, macKey) = deriveKeys(passphrase, salt)
        val expected = hmac(macKey, salt + nonce + ciphertext)
        if (!constantTimeEquals(mac, expected)) {
            throw BadPasswordException("Wrong passphrase or corrupt payload.")
        }
        val plain = xorKeystream(ciphertext, encKey, nonce)
        return String(plain, Charsets.UTF_8)
    }

    // ---- primitives ----

    private fun deriveKeys(passphrase: String, salt: ByteArray): Pair<ByteArray, ByteArray> {
        val spec = PBEKeySpec(passphrase.toCharArray(), salt, ITERATIONS, 40 * 8)
        val dk = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1").generateSecret(spec).encoded
        return dk.copyOfRange(0, 20) to dk.copyOfRange(20, 40)
    }

    private fun xorKeystream(input: ByteArray, encKey: ByteArray, nonce: ByteArray): ByteArray {
        val out = ByteArray(input.size)
        var produced = 0
        var counter = 0
        while (produced < input.size) {
            val block = hmac(encKey, nonce + be32(counter))
            var i = 0
            while (i < block.size && produced < input.size) {
                out[produced] = (input[produced].toInt() xor block[i].toInt()).toByte()
                produced++; i++
            }
            counter++
        }
        return out
    }

    private fun hmac(key: ByteArray, message: ByteArray): ByteArray {
        val mac = Mac.getInstance(HMAC)
        mac.init(SecretKeySpec(key, HMAC))
        return mac.doFinal(message)
    }

    private fun be32(v: Int): ByteArray = byteArrayOf(
        ((v ushr 24) and 0xFF).toByte(),
        ((v ushr 16) and 0xFF).toByte(),
        ((v ushr 8) and 0xFF).toByte(),
        (v and 0xFF).toByte()
    )

    private fun constantTimeEquals(a: ByteArray, b: ByteArray): Boolean {
        if (a.size != b.size) return false
        var diff = 0
        for (i in a.indices) diff = diff or (a[i].toInt() xor b[i].toInt())
        return diff == 0
    }
}

File: companion-ios/SentryKey/Sync/SyncCrypto.swift
import Foundation
import CommonCrypto

/// Optional passphrase encryption for the phone <-> watch BLE sync string.
///
/// Wire-compatible with the watch (`source/SyncCrypto.mc`) and Android
/// (`SyncCrypto.kt`); all three are pinned to the same canonical test vector.
/// Built on HMAC-SHA1 because the Garmin watch's native SHA-1 is unreliable, so
/// the watch only has a pure Monkey C HMAC-SHA1 to work from.
///
/// Wire format:
///   "SKENC1:" + base64( salt(16) || nonce(8) || mac(20) || ciphertext )
///   key    = PBKDF2-HMAC-SHA1(passphrase, salt, ITERATIONS, dkLen=40)
///   encKey = key[0..20), macKey = key[20..40)
///   keystream block i = HMAC-SHA1(encKey, nonce || be32(i))
///   ciphertext = plaintext XOR keystream
///   mac    = HMAC-SHA1(macKey, salt || nonce || ciphertext)
enum SyncCrypto {
    static let marker = "SKENC1:"
    static let iterations = 1000   // must match SyncCrypto.kt / SyncCrypto.mc

    private static let saltLen = 16
    private static let nonceLen = 8
    private static let macLen = 20

    static func isEncrypted(_ payload: String) -> Bool { payload.hasPrefix(marker) }

    static func encrypt(_ plaintext: String, passphrase: String) -> String {
        var salt = Data(count: saltLen)
        var nonce = Data(count: nonceLen)
        _ = salt.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, saltLen, $0.baseAddress!) }
        _ = nonce.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, nonceLen, $0.baseAddress!) }
        return encryptWith(plaintext, passphrase: passphrase, salt: salt, nonce: nonce)
    }

    /// Deterministic core, separated so a test can pin a fixed salt/nonce vector.
    static func encryptWith(_ plaintext: String, passphrase: String, salt: Data, nonce: Data) -> String {
        let (encKey, macKey) = deriveKeys(passphrase: passphrase, salt: salt)
        let ciphertext = keystreamXor(Data(plaintext.utf8), encKey: encKey, nonce: nonce)
        let mac = hmacSha1(key: macKey, message: salt + nonce + ciphertext)
        let blob = salt + nonce + mac + ciphertext
        return marker + blob.base64EncodedString()
    }

    /// Returns the plaintext sync string, or nil if the passphrase is wrong /
    /// the payload is corrupt (MAC mismatch).
    static func decrypt(_ payload: String, passphrase: String) -> String? {
        guard isEncrypted(payload), !passphrase.isEmpty else { return nil }
        let b64 = String(payload.dropFirst(marker.count))
        guard let blob = Data(base64Encoded: b64), blob.count >= saltLen + nonceLen + macLen else { return nil }

        let salt = blob.subdata(in: 0..<saltLen)
        let nonce = blob.subdata(in: saltLen..<(saltLen + nonceLen))
        let mac = blob.subdata(in: (saltLen + nonceLen)..<(saltLen + nonceLen + macLen))
        let ciphertext = blob.subdata(in: (saltLen + nonceLen + macLen)..<blob.count)

        let (encKey, macKey) = deriveKeys(passphrase: passphrase, salt: salt)
        let expected = hmacSha1(key: macKey, message: salt + nonce + ciphertext)
        guard constantTimeEquals(mac, expected) else { return nil }

        let plain = keystreamXor(ciphertext, encKey: encKey, nonce: nonce)
        return String(decoding: plain, as: UTF8.self)
    }

    // MARK: - primitives

    private static func deriveKeys(passphrase: String, salt: Data) -> (Data, Data) {
        let dk = pbkdf2(passphrase: passphrase, salt: salt, iterations: iterations, keyLen: 40)
        return (dk.subdata(in: 0..<20), dk.subdata(in: 20..<40))
    }

    private static func pbkdf2(passphrase: String, salt: Data, iterations: Int, keyLen: Int) -> Data {
        var derived = Data(count: keyLen)
        let pwData = Data(passphrase.utf8)
        _ = derived.withUnsafeMutableBytes { derivedPtr in
            salt.withUnsafeBytes { saltPtr in
                pwData.withUnsafeBytes { pwPtr in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        pwPtr.bindMemory(to: Int8.self).baseAddress, pwData.count,
                        saltPtr.bindMemory(to: UInt8.self).baseAddress, salt.count,
                        CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA1),
                        UInt32(iterations),
                        derivedPtr.bindMemory(to: UInt8.self).baseAddress, keyLen
                    )
                }
            }
        }
        return derived
    }

    private static func keystreamXor(_ input: Data, encKey: Data, nonce: Data) -> Data {
        var out = Data(count: input.count)
        var produced = 0
        var counter: UInt32 = 0
        while produced < input.count {
            let block = hmacSha1(key: encKey, message: nonce + be32(counter))
            var i = 0
            while i < block.count && produced < input.count {
                out[produced] = input[produced] ^ block[i]
                produced += 1; i += 1
            }
            counter += 1
        }
        return out
    }

    private static func hmacSha1(key: Data, message: Data) -> Data {
        var out = Data(count: Int(CC_SHA1_DIGEST_LENGTH))
        out.withUnsafeMutableBytes { outPtr in
            key.withUnsafeBytes { keyPtr in
                message.withUnsafeBytes { msgPtr in
                    CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA1),
                           keyPtr.baseAddress, key.count,
                           msgPtr.baseAddress, message.count,
                           outPtr.baseAddress)
                }
            }
        }
        return out
    }

    private static func be32(_ v: UInt32) -> Data {
        Data([UInt8((v >> 24) & 0xFF), UInt8((v >> 16) & 0xFF), UInt8((v >> 8) & 0xFF), UInt8(v & 0xFF)])
    }

    private static func constantTimeEquals(_ a: Data, _ b: Data) -> Bool {
        guard a.count == b.count else { return false }
        var diff: UInt8 = 0
        for i in 0..<a.count { diff |= a[i] ^ b[i] }
        return diff == 0
    }
}
